signwriting_animation/diffusion/tests/test_overfitting.py

```diff
@@ -0,0 +1,118 @@
+    pil_img = signwriting_to_clip_image(signwriting_str)
+    sw_img = clip_processor(images=pil_img, return_tensors="pt").pixel_values.to(device)
+    return x, timesteps, past_motion, sw_img
+
+def main():
+    # Device and processor setup
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+    clip_processor = CLIPProcessor.from_pretrained("openai/clip-vit-base-patch32")
+
+    # Model hyperparameters
+    num_keypoints = 21
+    num_dims_per_keypoint = 3
+    num_past_frames = 10
+
+    # Build model (ensure dropout=0 and no randomness)
+    model = SignWritingToPoseDiffusion(
+        num_keypoints=num_keypoints,
+        num_dims_per_keypoint=num_dims_per_keypoint,
+        embedding_arch="openai/clip-vit-base-patch32",
+        num_latent_dims=32,   
+        ff_size=64,
+        num_layers=2,
+        num_heads=2,
+        dropout=0.0,
+        cond_mask_prob=0
+    ).to(device)
+    optimizer = optim.Adam(model.parameters(), lr=1e-4)
+    loss_fn = torch.nn.MSELoss()
+
+    # Prepare two overfit samples: all zeros, all ones
+    sample_configs = [
+        # All zeros
+        ("AS14c20S27106M518x529S14c20481x471S27106503x489", 0.0),
+        # All ones
+        ("AS18701S1870aS2e734S20500M518x533S1870a489x515S18701482x490S20500508x496S2e734500x468", 1.0)
+    ]
+    samples = []
+    for sw, val in sample_configs:
+        x, timesteps, past_motion, sw_img = make_sample(
+            sw, val, device, clip_processor,
+            num_past_frames=num_past_frames,
+            num_keypoints=num_keypoints,
+            num_dims=num_dims_per_keypoint)
+        samples.append((x, timesteps, past_motion, sw_img, val))
+
+
+    # Training loop
+    num_epochs = 2000
+    losses = []
+    model.train()
+    for epoch in range(num_epochs):
+        epoch_loss = 0
+        for x, timesteps, past_motion, sw_img, val in samples:
+            optimizer.zero_grad()
+            output = model(x, timesteps, past_motion, sw_img)
+            target = torch.full_like(output, val)
+            loss = loss_fn(output, target)
+            loss.backward()
+            optimizer.step()
+            epoch_loss += loss.item()
+        avg_loss = epoch_loss / len(samples)
+        losses.append(avg_loss)
+        if epoch % 50 == 0:
+            print(f"Epoch {epoch}, Loss: {avg_loss:.6f}")
+        if epoch % 200 == 0:
+            print("Output min/max:", output.min().item(), output.max().item())
+
+    # Plot and save the training loss curve
+    plt.figure()
+    plt.plot(losses)
+    plt.xlabel("Epoch")
+    plt.ylabel("Loss")
+    plt.title("Overfit Loss Curve")
+    plt.tight_layout()
+    plt.savefig("overfit_loss_curve.png")
+    plt.close()
+    print("Saved loss curve to overfit_loss_curve.png")
+
+    model.eval()
+    with torch.no_grad():
+        for idx, (x, timesteps, past_motion, sw_img, val) in enumerate(samples):
+            output = model(x, timesteps, past_motion, sw_img)
```

Comment:
so, this is "kinda" correct - indeed, when asking the model for a single step, it can predict the result. 
But, the correct way to sample from a diffusion model is using https://github.com/AIGAnimation/CAMDM/blob/0c744b84133ca453dcf45aafdfbf49c6ff2841f9/PyTorch/diffusion/gaussian_diffusion.py#L626

Like Rong's work for example (her code is still quite messy though):
https://github.com/sign-language-processing/fluent-pose-synthesis/pull/13/files#diff-13f132104203d37aae4eb9e5c7de469b50a07bd55b5be6a30b02ebe6f385724dR319